src/java/org/apache/cassandra/db/compaction/UnifiedCompactionStrategy.java

```diff
@@ -1094,8 +1095,17 @@ Collection<CompactionAggregate.UnifiedAggregate> getCompactionAggregates(Arena a
                                                                       unbucketed::addAll);
 
             List<CompactionAggregate.UnifiedAggregate> aggregates = new ArrayList<>();
+            Set<CompactionSSTable> selectedSSTables = new HashSet<>();
             for (Bucket bucket : buckets)
-                aggregates.add(bucket.constructAggregate(controller, spaceAvailable, arena));
+            {
+                CompactionAggregate.UnifiedAggregate aggregate = bucket.constructAggregate(controller, spaceAvailable, arena);
+                // if some sstables are already selected, skip the bucket and aggregate
+                if (aggregate.getSelected().sstables().stream().anyMatch(selectedSSTables::contains))
```

Comment:
I didn't do the skipping inside `assignOverlapsIntoBuckets` because `UnifiedAggregate#selected` is only subset of bucket's sstables. Hopefully, it'd skip less aggregates compared to skipping inside `assignOverlapsIntoBuckets` with all sstables from each bucket.

Reply:
A proper solution to this involves rewriting `assignOverlapsIntoBuckets` for `InclusionMethod`s `SINGLE` and `NONE`. It should walk the sections list in size (the order perhaps constructed by a `SortingIterator` of indexes), but when it selects an index it should use the in-order section list to:
- expand left and right while the sections overlap with the selected section (to expand the bucket for `SINGLE`)
- walk left and right while the sections overlap with the expanded set and mark all as used
The selection should then skip over sections marked as used, and proceed until the size becomes smaller than the threshold.

Reply:
I have something like this in mind:
```
Index: src/java/org/apache/cassandra/utils/Overlaps.java
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/src/java/org/apache/cassandra/utils/Overlaps.java b/src/java/org/apache/cassandra/utils/Overlaps.java
--- a/src/java/org/apache/cassandra/utils/Overlaps.java	(revision 584642bcf0f543e0616359b154aa541a98157520)
+++ b/src/java/org/apache/cassandra/utils/Overlaps.java	(date 1727961328245)
@@ -19,6 +19,7 @@
 package org.apache.cassandra.utils;
 
 import java.util.ArrayList;
+import java.util.BitSet;
 import java.util.Collection;
 import java.util.Comparator;
 import java.util.HashSet;
@@ -27,6 +28,7 @@
 import java.util.Set;
 import java.util.function.BiPredicate;
 import java.util.function.Consumer;
+import java.util.stream.IntStream;
 
 public class Overlaps
 {
@@ -123,6 +125,104 @@
                                                            List<Set<E>> overlaps,
                                                            BucketMaker<E, B> bucketer,
                                                            Consumer<Set<E>> unselectedHandler)
+    {
+        switch (inclusionMethod)
+        {
+            case TRANSITIVE:
+                return assignOverlapsTransitive(threshold, overlaps, bucketer, unselectedHandler);
+            case SINGLE:
+            case NONE:
+                return assignOverlapsSingleOrNone(threshold, inclusionMethod, overlaps, bucketer, unselectedHandler);
+            default:
+                throw new AssertionError();
+        }
+    }
+
+    private static <E, B> List<B> assignOverlapsSingleOrNone(int threshold,
+                                                             InclusionMethod inclusionMethod,
+                                                             List<Set<E>> overlaps,
+                                                             BucketMaker<E, B> bucketer,
+                                                             Consumer<Set<E>> unselectedHandler)
+    {
+        List<B> buckets = new ArrayList<>();
+        int regionCount = overlaps.size();
+        SortingIterator<Integer> bySize = new SortingIterator<>((a, b) -> Integer.compare(overlaps.get(b).size(),
+                                                                                          overlaps.get(a).size()),
+                                                                IntStream.range(0, overlaps.size()).boxed().toArray());
+
+        BitSet used = new BitSet(overlaps.size());
+        while (bySize.hasNext())
+        {
+            final int i = bySize.next();
+            if (used.get(i))
+                continue;
+
+            Set<E> bucket = overlaps.get(i);
+            if (bucket.size() < threshold)
+                break;  // no more buckets will be above threshold
+            used.set(i);
+
+            Set<E> allOverlapping = bucket;
+            int j = i - 1;
+            int k = i + 1;
+            int startIndex = i;
+            int endIndex = i + 1;
+            if (inclusionMethod == InclusionMethod.SINGLE)
+            {
+                // expand the bucket to include all overlapping sets
+                allOverlapping = new HashSet<>(bucket);
+                Set<E> overlapTarget = bucket;
+                for (; j >= 0 && !used.get(j); --j)
+                {
+                    Set<E> next = overlaps.get(j);
+                    if (!setsIntersect(next, overlapTarget))
+                        break;
+                    allOverlapping.addAll(next);
+                    used.set(j);
+                }
+                startIndex = j + 1;
+                for (; k < regionCount && !used.get(k); ++k)
+                {
+                    Set<E> next = overlaps.get(k);
+                    if (!setsIntersect(next, overlapTarget))
+                        break;
+                    allOverlapping.addAll(next);
+                    used.set(k);
+                }
+                endIndex = k;
+            }
+            // Now mark all overlapping with the extended as used
+            Set<E> overlapTarget = allOverlapping;
+            for (; j >= 0 && !used.get(j); --j)
+            {
+                Set<E> next = overlaps.get(j);
+                if (!setsIntersect(next, overlapTarget))
+                    break;
+                used.set(j);
+                unselectedHandler.accept(next);
+            }
+            for (; k < regionCount && !used.get(k); ++k)
+            {
+                Set<E> next = overlaps.get(k);
+                if (!setsIntersect(next, overlapTarget))
+                    break;
+                used.set(k);
+                unselectedHandler.accept(next);
+            }
+            buckets.add(bucketer.makeBucket(overlaps, startIndex, endIndex));
+        }
+        for (int i = 0; i < regionCount; ++i)
+        {
+            if (!used.get(i))
+                unselectedHandler.accept(overlaps.get(i));
+        }
+        return buckets;
+    }
+
+    private static <E, B> List<B> assignOverlapsTransitive(int threshold,
+                                                           List<Set<E>> overlaps,
+                                                           BucketMaker<E, B> bucketer,
+                                                           Consumer<Set<E>> unselectedHandler)
     {
         List<B> buckets = new ArrayList<>();
         int regionCount = overlaps.size();
@@ -136,31 +236,27 @@
             int startIndex = i;
             int endIndex = i + 1;
 
-            if (inclusionMethod != InclusionMethod.NONE)
-            {
-                Set<E> allOverlapping = new HashSet<>(bucket);
-                Set<E> overlapTarget = inclusionMethod == InclusionMethod.TRANSITIVE
-                                       ? allOverlapping
-                                       : bucket;
-                int j;
-                for (j = i - 1; j >= lastEnd; --j)
-                {
-                    Set<E> next = overlaps.get(j);
-                    if (!setsIntersect(next, overlapTarget))
-                        break;
-                    allOverlapping.addAll(next);
-                }
-                startIndex = j + 1;
-                for (j = i + 1; j < regionCount; ++j)
-                {
-                    Set<E> next = overlaps.get(j);
-                    if (!setsIntersect(next, overlapTarget))
-                        break;
-                    allOverlapping.addAll(next);
-                }
-                i = j - 1;
-                endIndex = j;
-            }
+            Set<E> allOverlapping = new HashSet<>(bucket);
+            Set<E> overlapTarget = allOverlapping;
+            int j;
+            for (j = i - 1; j >= lastEnd; --j)
+            {
+                Set<E> next = overlaps.get(j);
+                if (!setsIntersect(next, overlapTarget))
+                    break;
+                allOverlapping.addAll(next);
+            }
+            startIndex = j + 1;
+            for (j = i + 1; j < regionCount; ++j)
+            {
+                Set<E> next = overlaps.get(j);
+                if (!setsIntersect(next, overlapTarget))
+                    break;
+                allOverlapping.addAll(next);
+            }
+            i = j - 1;
+            endIndex = j;
+
             buckets.add(bucketer.makeBucket(overlaps, startIndex, endIndex));
             for (int k = lastEnd; k < startIndex; ++k)
                 unselectedHandler.accept(overlaps.get(k));
```

Reply:
thanks for the changes, included in 3970b2c3431bfe28458b590e40622ad328f90105

Reply:
fixed SortingIterating construction when overlaps are empty https://github.com/datastax/cassandra/pull/1320/commits/ee019a3075b451eb9b8b1bccc459c0cb1d776a87